src/core/server/logging/appenders/rewrite/rewrite_appender.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { LogRecord, Appender, DisposableAppender } from '@kbn/logging';
+import {
+  createRewritePolicy,
+  rewritePolicyConfigSchema,
+  RewritePolicy,
+  RewritePolicyConfig,
+} from './policies';
+
+export interface RewriteAppenderConfig {
+  type: 'rewrite';
+  /**
+   * The {@link Appender | appender(s)} to pass the log event to after
+   * implementing the specified rewrite policy.
+   */
+  appenders: string[];
+  /**
+   * The {@link RewritePolicy | policy} to use to manipulate the provided data.
+   */
+  policy: RewritePolicyConfig;
+}
+
+/**
+ * Appender that can modify the `LogRecord` instances it receives before passing
+ * them along to another {@link Appender}.
+ * @internal
+ */
+export class RewriteAppender implements DisposableAppender {
+  public static configSchema = schema.object({
+    type: schema.literal('rewrite'),
+    appenders: schema.arrayOf(schema.string(), { defaultValue: [] }),
+    policy: rewritePolicyConfigSchema,
+  });
+
+  private appenders: Map<string, Appender> = new Map();
+  private readonly policy: RewritePolicy;
+
+  constructor(private readonly config: RewriteAppenderConfig) {
+    this.policy = createRewritePolicy(config.policy);
+  }
+
+  /**
+   * List of appenders that are dependencies of this appender.
+   *
+   * `addAppender` will throw an error when called with an appender
+   * reference that isn't in this list.
+   */
+  public get appenderRefs() {
+    return this.config.appenders;
+  }
+
+  /**
+   * Appenders can be "attached" to this one so that the RewriteAppender
+   * is able to act as a sort of middleware by calling `append` on other appenders.
+   *
+   * As appenders cannot be attached to each other until they are created,
+   * the `addAppender` method is used to pass in a configured appender.
+   */
+  public addAppender(appenderRef: string, appender: Appender) {
+    if (!this.appenderRefs.includes(appenderRef)) {
+      throw new Error(
+        `addAppender was called with an appender key that is missing from the appenderRefs: "${appenderRef}".`
+      );
+    }
+
```

Comment:
Too early for it probably, but we might want a `MiddlewareAppender` base / abstract class at some point in the future if we're going to add additional 'middlewares' to handle this `addAppender` logic, and most of the implementation of `append` (the nested appender looping)

Reply:
Log4j has something like this internally called `AttachableAppender` - Basically an appender that you can attach other appenders to. I thought about doing the same thing here but wasn't sure if it was worth introducing a new concept since we only have one concrete use case for it ATM.

Open to adding it though; WDYT?

Reply:
Let's wait for more cases before introducing a new abstraction.